File: nim/day14.nim
import sets, day10


type Coordinate = tuple[x, y: int]
const
  instructions = "hwlqcszp-"
  deltas: array[4, Coordinate] = [(1, 0), (-1, 0), (0, 1), (0, -1)]


func createMaze(): HashSet[Coordinate] =
  for row in 0 .. 127:
    let
      word = instructions & $row
      binHash = knotHashing(word, binOut = true)
    for i, n in binHash:
      if n == '1':
        result.incl((row, i))

var maze = createMaze()
echo maze.len


proc dfs(coord: Coordinate) =
  for delta in deltas:
    let candidate = (coord.x + delta.x, coord.y + delta.y)
    if candidate in maze:
      maze.excl(candidate)
      dfs(candidate)


var regions: int
while maze.len > 0:
  dfs(maze.pop())
  inc regions

echo regions

File: python/day14.py
from day10 import knot_hashing


instructions = 'hwlqcszp'
maze = set()

for row in range(128):
    word = f'{instructions}-{row}'
    hex_hash = knot_hashing(word)
    bin_hash = f'{int(hex_hash, 16):0128b}'
    for i, n in enumerate(bin_hash):
        if n == '1':
            maze.add((row, i))

print(len(maze))



def dfs(start):
    stack = [start]
    while stack:
        (x, y) = stack.pop()
        for dx, dy in DELTAS:
            candidate = x+dx, y+dy
            if candidate in maze:
                stack.append(candidate)
                maze.remove(candidate)


DELTAS = ((1, 0), (-1, 0), (0, 1), (0, -1))
regions = 0

while maze:
    dfs(maze.pop())
    regions += 1

print(regions)
